README
= validates_date_time

If you find this plugin useful, please consider a donation to show your support!

  http://www.paypal.com/cgi-bin/webscr?cmd=_send-money
  
  Email address: jonathan.viney@gmail.com
  
== Instructions

This plugin adds the ability to do stricter date and time checking with ActiveRecord.

The validators can be used to parse strings into <tt>Date</tt> and <tt>Time</tt> objects as well as restrict
an attribute based on other dates or times.

  class Person < ActiveRecord::Base
    validates_date     :date_of_birth
    validates_time     :time_of_birth
    validates_date_time :date_and_time_of_birth
  end
  
Use <tt>:allow_nil</tt> to allow the value to be blank.

  class Person < ActiveRecord::Base
    validates_date :date_of_birth, :allow_nil => true
  end
  
== Installation
 
  script/plugin install http://svn.viney.net.nz/things/rails/plugins/validates_date_time 

== Supported formats

The default for the plugin is to expect dates in day/month/year format. If you are in the
US, you will want to change the default to month/day/year by placing the following in config/environment.rb

  ValidatesDateTime.us_date_format = true
  
Date format examples:
 
* 2006-01-01
* 1 Jan 06
* 1 Jan 2006
* 10/1/06
* 1/1/2006
  
Time format examples:
* 1pm
* 10:11
* 12:30pm
* 8am

Datetime format examples:

* 1 Jan 2006 2pm
* 31/1/06 8:30am

== Examples

If an attribute value can not be parsed correctly, an error is added: 

  p = Person.new
  p.date_of_birth = "1 Jan 2006"
  p.time_of_birth = "5am"
  p.save # true
  
  p.date_of_birth = "30 Feb 2006"
  p.save # false, 30 feb is invalid for obvious reasons
   
  p.date_of_birth = "java is better than ruby"
  p.save # false
  
In the final example, as I'm sure you are aware, the record failed to save not only
because "java is better than ruby" is an invalid date, but more importantly, because the statement is blatantly false. ;) 

=== Restricting date and time ranges

Using the <tt>:before</tt> and <tt>:after</tt> options you can restrict a date or time value based on other attribute values
and predefined values. You can pass as many value to :before or :after as you like.

  class Person
    validates_date :date_of_birth, :before => [:date_of_death, Proc.new { 1.day.from_now_to_date}], :after => '1 Jan 1900'
    validates_date :date_of_death, :before => Proc.new { 1.day.from_now.to_date }
  end
  
  p = Person.new
  p.date_of_birth = '1800-01-01'
  p.save  # false
  p.errors[:date_of_birth] # must be after 1 Jan 1900
  
  p.date_of_death = Date.new(2010, 1, 1)
  p.save  # false
  p.errors[:date_of_death] # must be before <1 day from now>
  
  p.date_of_birth = '1960-03-02'
  p.date_of_death = '2003-06-07'
  p.save  # true
  
You can customise the error messages for dates or times that fall outside the required range. The boundary date will be substituted in for %s. Eg:

  class Person
    validates_date :date_of_birth, :after => Date.new(1900, 1, 1), :before => Proc.new { 1.day.from_now.to_date }, :before_message => 'Ensure it is before %s', :after_message => 'Ensure it is before %s'
  end
 
Suggestions, comments, problems are all welcome. You'll find me at jonathan.viney@gmail.com


CHANGELOG
[2012-02-22]

* Add support for Rails 3.0 [Kornelius Kalnbach]

[5 March 2009]

* Add support for I18N [Roman Shterenzon]

[18 February 2009]

* Fix errors caused by assigning values that don't respond to #strip.

[26 May 2008]

* Fix incorrect validation of 24 hour times including meridian, eg: 18:20 AM.

[2 May 2008]

* Permenently remove old ActiveRecord::Validations::DateTime namespace.

[19 March 2008]

* Fix README typo.

[11 February 2008]

* Make :after add an error if the date is the same as the restriction.

[7 January 2008]

* Add support for ISO8601 formats.

[14 December 2007]

* Allow 1 through 6 digits for microseconds.

[11 December 2007]

* Parse microseconds from time and datetime strings.

[28 September 2007]

* Preserve ActiveRecord::Validations::DateTime.us_date_format for compatibility.

[27 September 2007]

* Improve the handling of multiparameter attributes.

* BACKWARDS INCOMPATIBLE: Change namespace from ActiveRecord::Validations::DateTime to ValidatesDateTime.

[19 September 2007]

* Almost a complete rewrite! API unchanged. Let me know if there are any issues.

[2 Sep 2007]

 * Allow date and time formats like "February 5, 2006" and "September 01, 2007 06:10". [Mark A. Lane]

[7 Jun 07]

 * Datetime restrictions should be parsed as datetimes [Adam Meehan]
 
[Earlier]

 * Remove unused colon separated date format (dd:mm:yy) from parsing


test/person.rb
class Person < ActiveRecord::Base
  def self.columns()
    @columns ||= []
  end
  
  def self.column(name, sql_type = nil, default = nil, null = true)
    columns << ActiveRecord::ConnectionAdapters::Column.new(name.to_s, default, sql_type.to_s, null)
  end

  column :name
  column :date_of_birth,   :date
  column :date_of_death,   :date
  column :date_of_visit,   :date
  
  column :date_of_arrival,   :date
  column :date_of_departure, :date
  
  column :time_of_birth, :time
  column :time_of_death, :time
  
  validates_presence_of :name
  
  validates_date :date_of_birth,   :if => Proc.new { |p| p.date_of_birth? }, :before => nil
  validates_date :date_of_death,   :if => Proc.new { |p| p.date_of_death? }, :before => Proc.new { 1.day.from_now.to_date }, :after => Proc.new { |p| p.date_of_birth }
  
  validates_date :date_of_arrival, :if => Proc.new { |p| p.date_of_arrival? }, :before => Proc.new { |p| p.date_of_departure }, :before_message => "avant %s", :after_message => "apres %s"
  
  validates_time :time_of_birth, :if => Proc.new { |p| p.time_of_birth? }
  
  # Want to be able to use update_attributes
  def save
    valid?
  end
end


test/schema.rb
ActiveRecord::Schema.define :version => 0 do
  create_table :people, :force => true do |t|
    t.column :date_of_birth,          :date
    t.column :date_of_death,          :date
    
    t.column :date_of_arrival,        :date
    t.column :date_of_departure,      :date
  
    t.column :time_of_birth,          :time
    t.column :time_of_death,          :time
    
    t.column :date_and_time_of_birth, :datetime
  end
end


test/date_test.rb
require File.dirname(__FILE__) + '/abstract_unit'

class DateTest < Test::Unit::TestCase
  def test_valid_when_nil
    assert p.update_attributes!(:date_of_birth => nil, :date_of_death => nil)
  end
  
  def test_date_required
    assert_invalid_and_errors_match /invalid/, :required_date => ""
  end
  
  # Test 1/1/06 format
  def test_first_format
    { '1/1/01'  => '2001-01-01', '29/10/2005' => '2005-10-29', '8\12\63' => '1963-12-08',
      '07/06/2006' => '2006-06-07', '11\1\06' => '2006-01-11', '10.6.05' => '2005-06-10' }.each do |value, result|
      assert_update_and_equal result, :date_of_birth => value
    end
  end
  
  # Test 1 Jan 06 and 1 January 06 formats
  def test_second_format
    { '19 Mar 60'    => '1960-03-19', '22 dec 1985'      => '1985-12-22',
      '24 August 00' => '2000-08-24', '25 December 1960' => '1960-12-25'}.each do |value, result|
      assert_update_and_equal result, :date_of_birth => value
    end
  end
  
  # Test February 4 2006 formats
  def test_third_format
    { 'february 4 06' => '2006-02-04', 'DECember 25 1850' => '1850-12-25', 'February 5, 2006' => '2006-02-05' }.each do |value, result|
      assert_update_and_equal result, :date_of_birth => value
    end
  end
  
  def test_iso_format
    { '2006-01-01' => '2006-01-01', '1900-04-22' => '1900-04-22',
      '2008-03-04T20:33:41' => '2008-03-04', '2006-1-1' => '2006-01-01' }.each do |value, result|
      assert_update_and_equal result, :date_of_birth => value
    end
  end
  
  def test_format_without_validation
    assert_equal Date.new(2005, 12, 11).to_s, Person.new(:date_of_birth => "11/12/05").date_of_birth.to_s
  end
  
  def test_invalid_formats
    ['aksjhdaksjhd', 'meow', 'chocolate',
     '221 jan 05', '21 JAN 001', '1 Jaw 00', '1 Febrarary 2003', '30/2/06',
     '1/2/3/4', '11/22/33', '10/10/990', '189 /1 /9', '12\ f m', 100].each do |value|
      assert_invalid_and_errors_match /invalid/,  :date_of_birth => value
    end
  end
  
  def test_date_objects
    assert_update_and_equal '1963-04-05', :date_of_birth => Date.new(1963, 4, 5)
  end
  
  def test_before_and_after
    p.update_attributes!(:date_of_death => '1950-01-01')
    assert_invalid_and_errors_match /before/, :date_of_death => (Date.today + 1).to_s
    assert_invalid_and_errors_match /before/, :date_of_death => Date.new(2030, 1, 1)
    
    p.update_attributes!(:date_of_birth => '1950-01-01', :date_of_death => nil)
    assert_invalid_and_errors_match /after/, :date_of_death => '1950-01-01'
    assert p.update_attributes!(:date_of_death => Date.new(1951, 1, 1))
  end
  
  def test_before_and_after_with_custom_message
    assert_invalid_and_errors_match /avant/, :date_of_arrival => 2.years.from_now, :date_of_departure => 2.years.ago
    assert_invalid_and_errors_match /apres/, :date_of_arrival => '1792-03-03'
  end

  def test_before_and_after_with_i18n
    original_load_path = I18n.load_path
    
    I18n.load_path = [File.join(File.dirname(__FILE__),'fixtures/en.yml')]
    I18n.reload!
    
    assert_invalid_and_errors_match /i18n_before/, :date_of_death => (Date.today + 1).to_s
    assert_invalid_and_errors_match /i18n_after/, :date_of_birth => (Date.today + 1).to_s
  ensure
    I18n.load_path = original_load_path
    I18n.reload!
  end
  
  def test_dates_with_unknown_year
    assert p.update_attributes!(:date_of_birth => '9999-12-11')
    assert p.update_attributes!(:date_of_birth => Date.new(9999, 1, 1))
  end
  
  def test_us_date_format
    with_us_date_format do
      {'1/31/06'  => '2006-01-31', '28 Feb 01'  => '2001-02-28',
       '10/10/80' => '1980-10-10', 'July 4 1960' => '1960-07-04',
       '2006-03-20' => '2006-03-20'}.each do |value, result|
        assert_update_and_equal result, :date_of_birth => value
      end
    end
  end
  
  def test_blank
    assert p.update_attributes!(:date_of_birth => " ")
    assert_nil p.date_of_birth
  end
  
  def test_conversion_of_restriction_result
    assert_invalid_and_errors_match /Date of birth/, :date_of_death => Date.new(2001, 1, 1), :date_of_birth => Date.new(2005, 1, 1)
  end
  
  def test_multi_parameter_attribute_assignment_with_valid_date
    assert_nothing_raised do
      p.update_attributes!('date_of_birth(1i)' => '2006', 'date_of_birth(2i)' => '2', 'date_of_birth(3i)' => '10')
    end
    
    assert_equal Date.new(2006, 2, 10), p.date_of_birth
  end
  
  def test_multi_parameter_attribute_assignment_with_invalid_date
    assert_nothing_raised do
      assert !p.update_attributes('date_of_birth(1i)' => '2006', 'date_of_birth(2i)' => '2', 'date_of_birth(3i)' => '30')
    end
    
    assert p.errors[:date_of_birth]
  end
  
  def test_incomplete_multi_parameter_attribute_assignment
    assert_nothing_raised do
      assert !p.update_attributes('date_of_birth(1i)' => '2006', 'date_of_birth(2i)' => '1', 'date_of_birth(3i)' => '')
    end
    
    assert p.errors[:date_of_birth]
  end
end


lib/restrictions.rb
module ActiveRecord
  module Validations
    module DateTime
      module Restrictions
        class RestrictionError< StandardError #:nodoc:
        end
        
        class << self
          [:date, :time, :date_time].each do |method|
            class_eval <<-END
              def relative_#{method}_restrictions(configuration)
                [:before, :after].collect do |option|
                  [configuration[option]].flatten.compact.collect do |item|
                    case item
                      when Symbol, Proc, #{method.to_s.camelize} then item
                      when String then Parser.parse_#{method}(item)
                      else raise RestrictionError, "\#{item.class}:\#{item} invalid. Use either a Proc, String, Symbol or #{method.to_s.camelize} object."
                    end
                  end
                end
              end
              
              def #{method}_meets_relative_restrictions(value, record, restrictions, method)
                restrictions = restrictions.select do |restriction|
                  begin
                    case restriction
                      when Symbol
                        value.send(method, record.send(restriction)) rescue false
                        
                      when Proc
                        result = restriction.call(record)
                        result = Parser.parse_#{method}(result) unless result.is_a?(#{method.to_s.camelize})
                        value.send(method, result)
                        
                      when Date, Time
                        value.send(method, restriction)
                        
                      else
                        raise
                    end
                  rescue
                    raise RestrictionError, "Invalid restriction \#{restriction.class}:\#{restriction}"
                  end
                end
                
                restrictions.collect do |r|
                  case r
                    when Proc;   r.call(record)
                    when Symbol; r.to_s.humanize
                    else;        r
                  end
                end.first
              end
            END
          end
          
          def date_before(value, record, restrictions)
            date_meets_relative_restrictions(value, record, restrictions, :>)
          end
          
          def date_after(value, record, restrictions)
            date_meets_relative_restrictions(value, record, restrictions, :<=)
          end
          
          def time_before(value, record, restrictions)
            time_meets_relative_restrictions(value, record, restrictions, :>)
          end
          
          def time_after(value, record, restrictions)
            time_meets_relative_restrictions(value, record, restrictions, :<=)
          end
          
          def date_time_before(value, record, restrictions)
            date_time_meets_relative_restrictions(value, record, restrictions, :>)
          end
          
          def date_time_after(value, record, restrictions)
            date_time_meets_relative_restrictions(value, record, restrictions, :<=)
          end
        end
      end
    end
  end
end


lib/validates_date.rb
module ActiveRecord::Validations::Date
  def self.append_features(base)
    super
    base.extend(ClassMethods)
  end
  
  module ClassMethods
    def validates_date(*attr_names)
      configuration = { :message => 'is an invalid date', :on => :save }
      configuration.update(attr_names.pop) if attr_names.last.is_a?(Hash)
      
      validates_each(attr_names, configuration) do |record, attr_name, value|
        if date = parse_date_string(record.send("#{attr_name}_before_type_cast"))
          record.send("#{attr_name}=", date)
        else
          record.errors.add(attr_name, configuration[:message])
        end
      end
    end
    
   private
    def parse_date_string(string)
      return if string.nil?
      
      string.strip!
      
      # 22/1/06
      if  string =~ /^(\d{1,2})[\\\/](\d{1,2})[\\\/](\d{2}|\d{4})$/
        string = "#{ unambiguous_year $3 }-#{$2}-#{$1}"
      
      # 22 Feb 06 or 1 jun 2001
      elsif string =~ /^(\d{1,2}) (\w{3}) (\d{2}|\d{4})$/
        string = "#{ unambiguous_year $3 }-#{ Date::ABBR_MONTHNAMES.index($2.capitalize) }-#{$1}"
      
      # 2006-01-01, ignored
      elsif string =~ /^\d{4}-\d{2}-\d{2}$/
      
      # Not a valid date string
      else
        return
      end
      
      date_array = string.split('-').collect { |s| s.to_i }
      Date.new(date_array[0], date_array[1], date_array[2]) rescue nil
    end
    
    # Extract a 4-digit year from a 2-digit year.
    # If the number is less than 20, assume year 20#{number}
    # otherwise use 19#{number}. Ignore if already 4 digits.
    #
    # Eg:
    #    10 => 2010, 60 => 1960, 00 => 2000, 1963 => 1963
    def unambiguous_year(year)
      year.length == 2 ? (year.to_i < 20 ? "20#{year}" : "19#{year}") : year
    end
  end
end


lib/validates_date_time.rb
require File.dirname(__FILE__) + '/parser'
require File.dirname(__FILE__) + '/multiparameter_attributes'

module ValidatesDateTime
  def self.included(base)
    base.class_eval do
      extend ClassMethods
      include MultiparameterAttributes
    end
  end

  mattr_accessor :us_date_format
  self.us_date_format = false

  DEFAULT_TEMPORAL_VALIDATION_OPTIONS = {
    :before_message => "must be before %s",
    :after_message  => "must be after %s"
  }.freeze

  class ValidatesDateTimeRestriction < Struct.new(:raw_value, :parse_method)
    def value(record)
      @last_value = case raw_value
        when Symbol
          record.send(raw_value)
        when Proc
          raw_value.call(record)
        else
          raw_value
      end

      @last_value = parse(@last_value)
    end

    def parse(string)
      ActiveRecord::ConnectionAdapters::Column.send("string_to_#{parse_method}", string)
    end

    def last_value
      @last_value
    end

    def to_s
      if raw_value.is_a?(Symbol)
        raw_value.to_s.humanize
      else
        @last_value.to_s
      end
    end
  end


  module ClassMethods
    def validates_date(*args)
      options = temporal_validation_options({ :message => "is an invalid date" }, args)
      attr_names = args

      # We must remove this from the configuration that is passed to validates_each because
      # we want to have our own definition of nil that uses the before_type_cast value
      allow_nil = options.delete(:allow_nil)

      prepare_restrictions(options, :date)

      validates_each(attr_names, options) do |record, attr_name, value|
        raw_value = record.send("#{attr_name}_before_type_cast")

        # If value that is unable to be parsed, and a blank value where allow_nil is not set are both invalid
        if (!raw_value.blank? and !value) || (raw_value.blank? and !allow_nil)
          record.errors.add(attr_name, options[:message])
        elsif value
          validate_before_and_after_restrictions(record, attr_name, value, options)
        end
      end
    end

    def validates_time(*args)
      options = temporal_validation_options({ :message => "is an invalid time" }, args)
      attr_names = args

      allow_nil = options.delete(:allow_nil)
      prepare_restrictions(options, :dummy_time)

      validates_each(attr_names, options) do |record, attr_name, value|
        raw_value = record.send("#{attr_name}_before_type_cast")

        if (!raw_value.blank? and !value) || (raw_value.blank? and !allow_nil)
          record.errors.add(attr_name, options[:message])
        elsif value
          validate_before_and_after_restrictions(record, attr_name, value, options)
        end
      end
    end

    def validates_date_time(*args)
      options = temporal_validation_options({ :message => "is an invalid date time" }, args)
      attr_names = args

      allow_nil = options.delete(:allow_nil)
      prepare_restrictions(options, :time)

      validates_each(attr_names, options) do |record, attr_name, value|
        raw_value = record.send("#{attr_name}_before_type_cast")

        if (!raw_value.blank? and !value) || (raw_value.blank? and !allow_nil)
          record.errors.add(attr_name, options[:message])
        elsif value
          validate_before_and_after_restrictions(record, attr_name, value, options)
        end
      end
    end

   private
    def validate_before_and_after_restrictions(record, attr_name, value, options)
      if options[:before]
        options[:before].each do |r|
          if r.value(record) and value >= r.last_value
            record.errors.add(attr_name, :before, :value => r, :message => options[:before_message] % r)
            break
          end
        end
      end

      if options[:after]
        options[:after].each do |r|
          if r.value(record) and value <= r.last_value
            record.errors.add(attr_name, :after, :value => r, :message => options[:after_message] % r)
            break
          end
        end
      end
    end

    def prepare_restrictions(options, parse_method)
      options[:before] = [*options[:before]].compact.map { |r| ValidatesDateTimeRestriction.new(r, parse_method) }
      options[:after] = [*options[:after]].compact.map { |r| ValidatesDateTimeRestriction.new(r, parse_method) }
    end

    def temporal_validation_options(options, args)
      options.reverse_merge!(DEFAULT_TEMPORAL_VALIDATION_OPTIONS)
      options.update(args.pop) if args.last.is_a?(Hash)
      options.assert_valid_keys :message, :before_message, :after_message, :before, :after, :if, :unless, :on, :allow_nil
      options
    end
  end
end

class ActiveRecord::Base
  include ValidatesDateTime
end


lib/multiparameter_attributes.rb
module ValidatesDateTime
  module MultiparameterAttributes
    def self.included(base)
      base.alias_method_chain :execute_callstack_for_multiparameter_attributes, :temporal_error_handling
    end
    
    def execute_callstack_for_multiparameter_attributes_with_temporal_error_handling(callstack)
      errors = []
      callstack.each do |name, values|
        klass = (self.class.reflect_on_aggregation(name.to_sym) || column_for_attribute(name)).klass
        
        if values.empty?
          send("#{name}=", nil)
        else
          column = column_for_attribute(name)
          
          if [:date, :time, :datetime].include?(column.type)
            values = values.map(&:to_s)
            
            result = case column.type
              when :date
                extract_date_from_multiparameter_attributes(values)
              when :time
                extract_time_from_multiparameter_attributes(values)
              when :datetime
	        date_values, time_values = values.slice!(0, 3), values
	        extract_date_from_multiparameter_attributes(date_values) + " " + extract_time_from_multiparameter_attributes(time_values)
            end
                   
            send("#{name}=", result)
          end
        end
      end
      unless errors.empty?
        raise ActiveRecord::MultiparameterAssignmentErrors.new(errors), "#{errors.size} error(s) on assignment of multiparameter attributes"
      end
    end
    
    def extract_date_from_multiparameter_attributes(values)
      [values[0], *values.slice(1, 2).map { |s| s.rjust(2, "0") }].join("-")
    end
    
    def extract_time_from_multiparameter_attributes(values)
      values.last(3).map { |s| s.rjust(2, "0") }.join(":")
    end
  end
end


lib/parser.rb
module ActiveRecord
  module ConnectionAdapters
    class Column
      class << self
        def string_to_date(value)
          return if value.blank?
          return value if value.is_a?(Date)
          return value.to_date if value.is_a?(Time) || value.is_a?(DateTime)
          
          year, month, day = case value.to_s.strip
            # 22/1/06, 22\1\06 or 22.1.06
            when /\A(\d{1,2})[\\\/\.-](\d{1,2})[\\\/\.-](\d{2}|\d{4})\Z/
              ValidatesDateTime.us_date_format ? [$3, $1, $2] : [$3, $2, $1]
            # 22 Feb 06 or 1 jun 2001
            when /\A(\d{1,2}) (\w{3,9}) (\d{2}|\d{4})\Z/
              [$3, $2, $1]
            # July 1 2005
            when /\A(\w{3,9}) (\d{1,2})\,? (\d{2}|\d{4})\Z/
              [$3, $1, $2]
            # 2006-01-01 OR 2006-1-1
            when /\A(\d{4})-(\d{1,2})-(\d{1,2})\Z/
              [$1, $2, $3]
            # 2006-01-01T10:10:10+13:00 OR 2006-1-1T10:10:10+13:00
            when /\A(\d{4})-(\d{1,2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\Z/
              [$1, $2, $3]
            # Not a valid date string
            else
              return nil
          end
          
          Date.new(unambiguous_year(year), month_index(month), day.to_i) rescue nil
        end
        
        def string_to_dummy_time(value)
          return if value.blank?
          return value if value.is_a?(Time) || value.is_a?(DateTime)
          return value.to_time(ActiveRecord::Base.default_timezone) if value.is_a?(Date)
          
          hour, minute, second, microsecond = case value.to_s.strip
            # 12 hours with minute and second
            when /\A(\d{1,2})[\. :](\d{2})[\. :](\d{2})\s?(am|pm)\Z/i
              [full_hour($1, $4), $2, $3]
            # 12 hour with minute: 7.30pm, 11:20am, 2 20PM
            when /\A(\d{1,2})[\. :](\d{2})\s?(am|pm)\Z/i
              return nil unless $1.to_i <= 12
              [full_hour($1, $3), $2]
            # 12 hour without minute: 2pm, 11Am, 7 pm
            when /\A(\d{1,2})\s?(am|pm)\Z/i
              [full_hour($1, $2)]
            # 24 hour: 22:30, 03.10, 12 30
            when /\A(\d{2})[\. :](\d{2})([\. :](\d{2})(\.(\d{1,6}))?)?\Z/
              [$1, $2, $4, $6]
            # 2006-01-01T10:10:10+13:00
            when /\A(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\Z/
              [$4, $5, $6]
            # Not a valid time string
            else
              return nil
          end
          
          Time.send(ActiveRecord::Base.default_timezone, 2000, 1, 1, hour.to_i, minute.to_i, second.to_i, microsecond.to_i) rescue nil
        end
        
        def string_to_time(value)
          return if value.blank?
          return value if value.is_a?(Time) || value.is_a?(DateTime)
          return value.to_time(ActiveRecord::Base.default_timezone) if value.is_a?(Date)
          
          value = value.to_s.strip
          
          if value =~ /\A(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\Z/
            time_array = [$1, $2, $3, $4, $5, $6].map!(&:to_i)
          else
            # The basic approach is to attempt to parse a date from the front of the string, splitting on spaces.
            # Once a date has been parsed, a time is extracted from the rest of the string.
            split_index = date = nil
            loop do
              split_index = value.index(' ', split_index ? split_index + 1 : 0)
              
              if split_index.nil? or date = string_to_date(value.first(split_index))
                break
              end
            end
            
            return if date.nil?
            
            time = string_to_dummy_time(value.last(value.size - split_index))
            return if time.nil?
            
            time_array = [date.year, date.month, date.day, time.hour, time.min, time.sec, time.usec]
          end
          
          # From schema_definitions.rb
          begin
            Time.send(ActiveRecord::Base.default_timezone, *time_array)
          rescue
            zone_offset = if Base.default_timezone == :local then DateTime.now.offset else 0 end
            # Append zero calendar reform start to account for dates skipped by calendar reform
            DateTime.new(*time_array[0..5] << zone_offset << 0) rescue nil
          end
        end
        
        def full_hour(hour, meridian)
          hour = hour.to_i
          if meridian.strip.downcase == 'am'
            hour == 12 ? 0 : hour
          else
            hour == 12 ? hour : hour + 12
          end
        end
        
        def month_index(month)
          return month.to_i if month.to_i.nonzero?
          Date::ABBR_MONTHNAMES.index(month.capitalize) || Date::MONTHNAMES.index(month.capitalize)
        end
        
        # Extract a 4-digit year from a 2-digit year.
        # If the number is less than 20, assume year 20#{number}
        # otherwise use 19#{number}. Ignore if already 4 digits.
        #
        # Eg:
        #    10 => 2010, 60 => 1960, 00 => 2000, 1963 => 1963
        def unambiguous_year(year)
          year = "#{year.to_i < 20 ? '20' : '19'}#{year}" if year.length == 2
          year.to_i
        end
      end
    end
  end
end
